rspace/src/main/scala/coop/rchain/rspace/merger/EventLogIndex.scala

```diff
@@ -0,0 +1,156 @@
+                    .whenA(touchPreStateJoin)
+              _ <- producesLinearRef
+                    .update(s => s + p)
+                    .unlessA(p.persistent)
+              _ <- producesPersistentRef
+                    .update(s => s + p)
+                    .whenA(p.persistent)
+            } yield ()
+          case c: Consume =>
+            for {
+              _ <- consumesLinearRef
+                    .update(s => s + c)
+                    .unlessA(c.persistent)
+              _ <- consumesPersistentRef
+                    .update(s => s + c)
+                    .whenA(c.persistent)
+            } yield ()
+          case c: COMM => {
+            for {
+              _ <- producesConsumedRef
+                    .update(s => s ++ c.produces)
+                    .whenA(c.peeks.isEmpty)
+              _ <- producesPeekedRef
+                    .update(s => s ++ c.produces)
+                    .whenA(c.peeks.nonEmpty)
+              _ <- consumesProducedRef
+                    .update(s => s + c.consume)
+            } yield ()
+          }
+        })
+      }
+      _ <- fs2.Stream.emits(ps).parJoinProcBounded.compile.drain
+
+      producesLinear         <- producesLinearRef.get
+      producesPersistent     <- producesPersistentRef.get
+      producesDestroyed      <- producesConsumedRef.get
+      producesPeeked         <- producesPeekedRef.get
+      consumesLinearAndPeeks <- consumesLinearRef.get
+      consumesPersistent     <- consumesPersistentRef.get
+      consumesCommed         <- consumesProducedRef.get
+      producesCopiedByPeek   <- producesCopiedByPeekRef.get
+      producesTouchingJoins  <- producesTouchingJoinsRef.get
+
+    } yield EventLogIndex(
+      producesLinear = producesLinear,
+      producesPersistent = producesPersistent,
+      producesConsumed = producesDestroyed,
+      producesPeeked = producesPeeked,
+      producesCopiedByPeek = producesCopiedByPeek,
+      producesTouchingBaseJoins = producesTouchingJoins,
+      consumesLinearAndPeeks = consumesLinearAndPeeks,
+      consumesPersistent = consumesPersistent,
+      consumesProduced = consumesCommed
+    )
+
+  implicit val monoid: Monoid[EventLogIndex] = new Monoid[EventLogIndex] {
+    override def empty: EventLogIndex =
+      EventLogIndex(
+        Set.empty,
+        Set.empty,
+        Set.empty,
+        Set.empty,
+        Set.empty,
+        Set.empty,
+        Set.empty,
+        Set.empty,
+        Set.empty
+      )
+
+    override def combine(x: EventLogIndex, y: EventLogIndex): EventLogIndex =
+      EventLogIndex.combine(x, y)
+  }
+
+  def combine(x: EventLogIndex, y: EventLogIndex): EventLogIndex = {
+
+    val newProducesCreated = (x, y).map(producesCreated).fold
+
+    /** if produce is copied by peek in one index and originated in another - it is considered as created in aggregate*/
+    val newProducesCopiedByPeek = (x, y).map(_.producesCopiedByPeek).fold diff newProducesCreated
```

Comment:
These two mappings and folds on a tuple are just operation only on the second value.
```scala
(x, y).map(f).fold == f(y)
```

Reply:
https://github.com/rchain/rchain/pull/3410/commits/e11ed628180c1950e81e0e6186a86994838a31f2